libraries/Microsoft.Bot.Builder/IBotTelemetryClientExtensions.cs

```diff
@@ -0,0 +1,26 @@
+﻿using System.Collections.Generic;
+
+namespace Microsoft.Bot.Builder
+{
+    public static class IBotTelemetryClientExtensions
+    {
+        /// <summary>
+        /// Adds the ability to call TrackPageView on the TelemetryClient if it implements IBotPageViewTelemetryClient as well as IBotTelemetryClient.
+        /// </summary>
+        /// <param name="telemetryClient">The TelemetryClient that implements IBotTelemetryClient.</param>
+        /// <param name="dialogName">The name of the dialog to log the entry / start for.</param>
+        /// <param name="properties">Named string values you can use to search and classify events.</param>
+        /// <param name="metrics">Measurements associated with this event.</param>
+        public static void TrackPageView(this IBotTelemetryClient telemetryClient, string dialogName, IDictionary<string, string> properties = null, IDictionary<string, double> metrics = null)
+        {
+            if (telemetryClient is IBotPageViewTelemetryClient pageViewClient)
+            {
+                pageViewClient.TrackPageView(dialogName, properties, metrics);
+            }
+            else
+            {
+                telemetryClient.TrackTrace("TelemetryClient cannot track PageView telemtry as it does not implement IBotPageViewTelemetryClient", Severity.Information, properties);
```

Comment:
can't we gently "fall back" to less sophisticated behavior (fro example, an event) rather than abruptly just trace saying "update your stuff."

I'm thinking the user of the telemetry functionality is not necessary the implementer of the telemetry provider. If the user is constrained for some reason to some particular telemetry provider (other than our appsignts solution) then they at least get some reasonable behavior. Or am I miss-representing the scenario?

Reply:
Surely if the user is using a different system then they will implement their own class that uses both interfaces?  In the case of our method we want to handle the situation where the page view is not logged because the interface has not been implemented.  I think a warning here is fine and it is probably an edge case.

The alternative would be to log additional fallback events using TrackEvent, but the danger then is that you start polluting someone's telemetry and this will only be in the case they have their own implementation. So, it feels better to warn them gracefully rather than explicitly log something.

Reply:
Given we shipped an interface, we implied there can be other implementations. So assuming there are, then those preexisting implementations don't have a TrackPageView concept the TrackEvent is the best thing that can happen.

I'm assuming a world where someone has preexisting telemetry and I just wanted to get the data into their telemetry rather than a warning message telling them that they must upgrade their telemetry capabilities.

In other words, the user of the telemetry client is in control, if they want the enhanced experience of pageview then they can upgrade their telemetry provider to get that. In the meantime they get an event.

(Having said that I don't have any personal experience of what it means to get a PageView rather than an Event in my telemetry. I'm assuming its better, but I'm also assuming an Event kinda works.)

Reply:
Updated Trace event to contain info as opposed to warning message as discussed.